### ISSUE TITLE
Serialisation of Panel Objects

### ISSUE THREAD

**Issue by pgierz** [holoviz/panel#1086]
Hi,

## Background

we are currently involved in a large federal project to run long climate simulations of the last 130,000 years including into the future. These runs will likely take a few months to complete on the supercomputer. On of the things I've been using panel for in personal use is to easily automate some of the visualisations I need to make. 

## Feature Request:
is it possible to serialise Panel objects? Let’s say I have something like this:

```
import panel as pn
pn.extension()

import numpy as np
import matplotlib.pyplot as plt 

data = np.random.random(100)
plot = plt.plot(data, 'bo')
panel_obj = pn.Row("# My plot is cool: ", plot)
```

What would be cool is you could now save panel_obj to some sort of file, and then open it again to get exactly the same contents back. If I understand correctly, the data is also indirectly saved into the panel object. I know there is panel_obj.save but how does one load again?

## Why?

If it were possible to embed some of the Panel (or, more generally all sorts of holoviz/bokeh stuff) into reusable files; I can image we could then get results as the model runs and then piece them together to get periodically updating snapshots of the simulations, plots that can be used for publications, slides, public outreach, and so on.

I've also raised this on the Discourse:
https://discourse.holoviz.org/t/serialising-panel-objects/58

**Comment by pgierz** [comment]
Hello again,

I now have a concrete use case: I've made a Trimesh with nice zooming and re-rendering of the triangular elements.

<img width="1008" alt="Screenshot 2020-02-25 at 09 27 26" src="https://user-images.githubusercontent.com/2444231/75228814-0bfa4380-57b1-11ea-890e-8c3fad9b73dc.png">

One side note: I can't get the mesh color to change, and the line width is a bit obstructive when you zoom too far out. Maybe some sort of dynamic control on the mesh's visibility depending on the zoom level might be nice; but I'd need help implementing that. 

Packing this into a panel pane also works easily. Saving does too:
```
pn_obj = pn.Pane(plot)
pn_obj.save("test.hvz")
```
This gives me a HTML. For many figures; this is probably entirely fine. The HTML also has save functionality as included in Bokeh; so end-users can use the button to create PNGs. I haven't seen what happens with widgets yet.

This brings to mind two (related) problems:
1) Providing a *full* export with all the data intact would be great; however, any widgets that dependent on external objects won't be able to function, since this would likely scale to packing up the entire notebook. Not really what I want.

2) Related to 1: If it is possible to determine that all the functionality to properly modify the figure (in this case, zooming and rerendering with data shader) is available; can we pickle it? Or something else that can be passed around to another session?

I'd be happy to get some points and try implementing something like this on my own; but admittedly would need **a lot** of guidance starting out.

**Comment by pgierz** [comment]
And, as a note:

```
with open("plot.hvz", "wb") as f:
    pickle.dump(plot, f)
```

gives:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-53-6f923421769f> in <module>
      1 with open("plot.hvz", "wb") as f:
----> 2     pickle.dump(plot, f)

AttributeError: Can't pickle local object 'HoloMap._dynamic_mul.<locals>.dynamic_mul'
```

Probably, this is what @jbednar was trying to explain to me on Discourse https://discourse.holoviz.org/t/serialising-panel-objects/58

**Comment by jbednar** [comment]
> Maybe some sort of dynamic control on the mesh's visibility depending on the zoom level might be nice; but I'd need help implementing that.

That should be easy enough to do if you follow the example at https://examples.pyviz.org/iex_trading/IEX_trading.html, which overlays a hoverable Element on top of a datashaded element, with data in it only when zoomed in enough to make that feasible.  Here you could choose an alpha value for the mesh that depends on zoom level, perhaps.

I'm not sure what the specific issue for dynamic_mul is, but yes, it's an example of the general issue that DynamicMap allows literally anything that Python can do (which is essentially anything that a computer can do!), and there's no way to ensure that _all_ of those possible things are serializable.  So it boils down to ensuring that you are only using things that are serializable. Here it sounds like you're looking for datashader-based HoloViews operations to be included in that set of serializable things, which is a reasonable request.  I would _think_ that it shouldn't be too difficult, but I've actually never tried it myself.

**Comment by philippjfr** [comment]
We're always going to need something like cloudpickle which can serialize inline functions. Without it I don't think there's any chance of serializing a DynamicMap.

**Comment by jbednar** [comment]
True. It looks like DynamicMap's implementation has a lot of unpicklable functions.  Some of those are just lambda functions, so they can simply be given names and put into the module namespace so that they can be unpickled. Others are locally scoped functions but that's just for convenience, so the same approach would work for them. But others appear to make use of the fact that they are locally scoped and acting as closures, which would be painful to rewrite.  @pgierz , maybe you could try https://github.com/cloudpipe/cloudpickle and see if that works for a DynamicMap? I'd assume cloudpickle objects are even less "archival" than regular pickle objects, though, even if it does work, i.e. that they would depend precisely on the version of Python used for creating them.

**Comment by pgierz** [comment]
Hello @jbednar, I had a look at cloud pickle. While I no longer get an Attribute error, I now have some thing else:

```

[20]:

import pickle
import cloudpickle
[22]:

with open("plot.hvz", "wb") as f:
    pickle.dump(plot, f)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-22-6f923421769f> in <module>
      1 with open("plot.hvz", "wb") as f:
----> 2     pickle.dump(plot, f)

AttributeError: Can't pickle local object 'HoloMap._dynamic_mul.<locals>.dynamic_mul'
[23]:

with open("plot.hvz", "wb") as f:
    cloudpickle.dump(plot, f)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-7216c1a3a31f> in <module>
      1 with open("plot.hvz", "wb") as f:
----> 2     cloudpickle.dump(plot, f)

~/anaconda3/envs/fesom_viz/lib/python3.8/site-packages/cloudpickle/cloudpickle_fast.py in dump(obj, file, protocol, buffer_callback)
     45     compatibility with older versions of Python.
     46     """
---> 47     CloudPickler(file, protocol=protocol, buffer_callback=buffer_callback).dump(obj)
     48 
     49 

~/anaconda3/envs/fesom_viz/lib/python3.8/site-packages/cloudpickle/cloudpickle_fast.py in dump(self, obj)
    544     def dump(self, obj):
    545         try:
--> 546             return Pickler.dump(self, obj)
    547         except RuntimeError as e:
    548             if "recursion" in e.args[0]:

TypeError: cannot pickle 'weakref' object
```
